Add partial implementation of accelerator API defined in #3493

stevew817 opened the pull request:
## Description
This PR implements partial support for delegating cryptographic operations to 'transparent' crypto accelerators, as defined in #3493. The operations implemented through this PR are `sign_hash`, `verify_hash` and `generate_key`, as requested in #3347. It also adds a basic 'test driver' implementation of those functions, as requested in #3344.

Currently, in order to compile in support for the test driver, `PSA_CRYPTO_DRIVER_TEST` needs to be defined. This is a temporary measure, since the contents of `psa_crypto_driver_wrappers.c` is envisioned to be auto-generated (ref. #3342). Once the auto-generation part is in place, the test driver can be compiled in/out by re-generating the wrapper file with/without the test driver description file.

## Status
**READY**

## Requires Backporting
NO  

## Migrations
NO

## Todos
- [X] Tests
- [X] Documentation
- [x] Changelog updated

## Steps to test or reproduce
Added corresponding (new) test suite `test_psa_crypto_drivers` to all.sh

stevew817 commented:
Rebased onto #3492 to pull in a small CI fix that got added there.

stevew817 commented:
Requesting review from @gilles-peskine-arm and @ronald-cron-arm

gilles-peskine-arm commented:
The gnutls failures in the “full configuration” job on Travis are happening on all PR today and we're investigating. On the other hand the `check_xxx` failures are on you.

stevew817 commented:
Force-push to rebase on latest changes from #3492

stevew817 commented:
rebased once again to the head of #3492

stevew817 commented:
Rebased on tip of development (now that #3492 is merged in) and cleaned up commit history.

@gilles-peskine-arm This PR is now ready for review and no longer pending preceding PR.

stevew817 commented on tests/include/drivers/signature.h:
```diff
@@ -0,0 +1,55 @@
+/*
+ * Test driver for signature functions
+ */
+/*  Copyright (C) 2020, ARM Limited, All Rights Reserved
```
Applied change to all new files

bensze01 replied:
Thanks!

stevew817 commented on tests/include/drivers/signature.h:
```diff
@@ -0,0 +1,55 @@
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  This file is part of mbed TLS (https://tls.mbed.org)
```
Applied change to all new files

stevew817 commented:
@bensze01 I've applied the copyright header changes to all source files that got created as part of this PR. Please re-approve.

gilles-peskine-arm commented on tests/src/drivers/signature.c:
```diff
@@ -0,0 +1,299 @@
+                                               signature + curve_bytes,
+                                               curve_bytes ) );
+cleanup:
+    /* There's no easy way to translate the error code except through a
+     * library function that's not exported. Use a debugger. */
```
`mbedtls_to_psa_error` can be used in test code.

gilles-peskine-arm commented on tests/src/drivers/signature.c:
```diff
@@ -0,0 +1,299 @@
+/*
+ * Test driver for signature functions
+ */
```
Please add a comment summarizing which features (entry points, algorithms, key types) are supported.

gilles-peskine-arm commented on tests/src/drivers/keygen.c:
```diff
@@ -0,0 +1,127 @@
+/*
+ * Test driver for signature functions
```
copypasta (in the `.h` as well)

gilles-peskine-arm commented on tests/include/drivers/test_driver.h:
```diff
@@ -0,0 +1,28 @@
+#ifndef PSA_CRYPTO_TEST_DRIVER_H
+#define PSA_CRYPTO_TEST_DRIVER_H
+
+#define PSA_CRYPTO_TEST_DRIVER_LIFETIME 0x7fffff
```
We'll have multiple test drivers, so it would be better to get into the habit of giving them some kind of name, even if it's only `TEST_DRIVER_1`.

stevew817 replied:
Are there really going to be more? Keep in mind this lifetime value is only relevant for the opaque drivers, since transparent doesn't use lifetime. It'll at some point also have to be merged into the autogenerated file, since I recall you wanted to have the location lifetime values declared in JSON?

gilles-peskine-arm replied:
We'll need to test builds with multiple drivers. At least some of the test drivers will have their expected location hard-coded so that they can check that they're being called with the correct location (dispatching to the wrong driver is a plausible mode of failure that we should test).

stevew817 replied:
I agree with wanting to test with multiple drivers, but don't you think that's a task for when testing with multiple drivers actually becomes feasible? Read: for when autogen is in place and defining a hardcoded location will actually be possible? This is just a placeholder...

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -1616,6 +1616,16 @@ component_test_se_default () {
 
+component_test_psa_crypto_drivers () {
+    msg "build: MBEDTLS_PSA_CRYPTO_DRIVERS w/ driver hooks"
+    scripts/config.py set MBEDTLS_PSA_CRYPTO_DRIVERS
```
Minor: I think it would make sense to enable `MBEDTLS_USE_PSA_CRYPTO`, so that TLS and X.509 get to invoke the transparent driver code. Not a blocker here, we can expand on the testing later.

stevew817 replied:
OK, I'll leave it for later.

gilles-peskine-arm commented on tests/src/drivers/keygen.c:
```diff
@@ -0,0 +1,127 @@
+
+    /* Copied from psa_crypto.c */
+#if defined(MBEDTLS_ECP_C)
+    if ( PSA_KEY_TYPE_IS_ECC( attributes->core.type ) && PSA_KEY_TYPE_IS_KEY_PAIR( attributes->core.type ) )
```
```suggestion
    if ( PSA_KEY_TYPE_IS_ECC_KEY_PAIR( psa_get_key_type( attributes ) )
```
Please avoid accessing fields of `psa_key_attributes_t` directly in test code.

ronald-cron-arm commented on tests/src/drivers/signature.c:
```diff
@@ -0,0 +1,299 @@
+
+    psa_status_t status = PSA_ERROR_NOT_SUPPORTED;
+
+#if defined(MBEDTLS_ECDSA_C) && defined(MBEDTLS_ECDSA_DETERMINISTIC) && \
```
My understanding of the code below is that it is a light version of `psa_sign_hash()` (didn't check completely though, but if it isn't I am not sure to understand why anyway).
The high-level structure of `psa_sign_hash()` is:
`psa_get_key_from_slot();`
`psa_driver_wrapper_sign_hash();`
`if not done do the operation with software`
Thus what about to reorganize in psa_crypto.c the "do the operation with software" in one function and just call this function here.
That way there should be less code duplication, thus less code maintenance and it makes clear that the code here is just the code in psa_crypto.c.

stevew817 replied:
That would mean exposing that "do the operation with software" function through a header, such that the test driver can reach it. I've had pushback before on this, since it can lead to confusion as to what the public API is.

Not saying I don't want to do it, but then it should be done consequently for all "do in software" operations.

gilles-peskine-arm replied:
We have a [process](https://github.com/ARMmbed/mbedtls/blob/development/docs/architecture/testing/invasive-testing.md#user-content-reliance-on-internal-details) for this: declare the internal function in a `library/*.h`, and if there's no other reason to make it non-`static`, use `MBEDTLS_STATIC_TESTABLE` instead of `static`.

ronald-cron-arm replied:
Yes it should be done consistently and the "do the operation with software" should be in an internal header thus currently in an header in library (not in include/psa).

stevew817 replied:
Since this will lead to further splintering of the psa_crypto.c core, can we delegate this work to a separate work unit?

ronald-cron-arm replied:
I would be good to get the whole picture for the PSA API as stake in this PR thus for generate_key(), sign_hash() and verify_hash(). Thus I would be in favor of doing this work in this PR. When we are ok with the overall result for those APIs we would generalized this in a separate PR.

stevew817 replied:
> This process is fairly new so there aren't many examples in the code base yet.
That's an understatement: There's exactly *one*. And it came in after I started this PR. Meaning *yet another* rebase. I'm going to forego the static declaration part for now and just declare the internal functions in an internal header. We previously established that the test drivers shouldn't be dependant on `MBEDTLS_TEST_HOOKS`, so the internal functions not being accessible unless that is defined is in issue regardless.

gilles-peskine-arm replied:
I'm personally ok with duplicating some code in the test drivers. It's code that uses stable interfaces so it's unlikely to break much, and we don't care all that much if there's a bug fix in the copied code that doesn't get applied to the copy in the test code. As long as these are clearly test drivers (as opposed to sample drivers), it's ok if they don't have the best hygiene and it's ok if they're buggy or even insecure in corner cases.

stevew817 replied:
The _internal functions that got split off when implementing driver delegation are now made `MBEDTLS_STATIC_TESTABLE`, which would make them available for basing an eventual fully-featured test driver off of them. Creating that fully featured test driver is however not in scope for this PR.

stevew817 replied:
OK, changed to unconditional external linkage. This may bloat code size for those not using LTO, though.

stevew817 replied:
> please remove your last commit

Only the last commit? I.e. you want to keep the `MBEDTLS_STATIC_TESTABLE` declarations? Or do you want me to revert the whole splitting up of the functions such that the software fallback is inlined in the `psa_crypto_xxx_hash()` again?

ronald-cron-arm replied:
I wrote this before you pushed the latest patch (following Gilles comment). Thus I propose to remove (rather than revert) the two last commits from this PR.

stevew817 replied:
Done.

stevew817 commented on library/psa_crypto_driver_wrappers.c:
```diff
@@ -0,0 +1,378 @@
+            drv->asymmetric->p_sign == NULL )
+        {
+            /* Key is defined in SE, but we have no way to exercise it */
+            return PSA_ERROR_INVALID_ARGUMENT;
```
Since `NOT_SUPPORTED` is used to indicate fallback to software implementation, and the software implementation wouldn't be able to do anything with a key stored inside an SE.

Would the alternative be to define another error value e.g. `PSA_ERROR_FALLBACK_REQUESTED`?

gilles-peskine-arm replied:
`NOT_SUPPORTED` indicates fallback only for transparent drivers. The core won't try to do anything with an opaque key.

gilles-peskine-arm replied:
> The core won't try to do anything with an opaque key.

￼Or rather shouldn't. I hadn't noticed that.

stevew817 commented on library/psa_crypto_driver_wrappers.c:
```diff
@@ -0,0 +1,378 @@
+                                           size_t signature_size,
+                                           size_t *signature_length )
+{
+#if defined(PSA_CRYPTO_DRIVER_PRESENT)
```
The point is to be able to `(void)` the function arguments in case there is nothing there to look at them, since otherwise the compiler will complain about unused variables.

stevew817 commented on library/psa_crypto_driver_wrappers.c:
```diff
@@ -0,0 +1,378 @@
+
+/* Include test driver definition when running tests */
+#if defined(PSA_CRYPTO_DRIVER_TEST)
+#undef PSA_CRYPTO_DRIVER_PRESENT
```
I was imagining each driver declaration contributing this block, and was trying to work around redefinition errors. Will change to a #ifndef instead of explicit redefinition.

gilles-peskine-arm replied:
```
#define FOO whatever
#define FOO whatever
```
is valid C as long as the two `whatever` are identical token sequences. A picky compiler could warn about this but as far as I recall we do have precedents of defining the same macro twice and nobody complained. (Somebody did complain once, but it turned out that the definitions didn't match in that particular configuration, which made the compiler's complaint perfectly justified.)

stevew817 replied:
Sure, but you pointed out someone could declare it with a value, whereas I'm only using the 'defined' property. Meaning I'll stick with the ifndef to make sure whatever value it gets declared with is retained.
Feel free to change when you get around to autogeneration.

gilles-peskine-arm commented on library/psa_crypto.c:
```diff
@@ -3659,24 +3656,18 @@ psa_status_t psa_sign_hash( psa_key_handle_t handle,
+    if( status != PSA_ERROR_NOT_SUPPORTED )
+        goto exit;
+
+    /* If the operation was not supported by any accelerator, try fallback. */
```
If the key is opaque, the code below this point attempts to access it as a transparent key. This only works if it's guaranteed that the code below isn't called on opaque keys, which in turn relies on the driver wrapper never returning `PSA_ERROR_NOT_SUPPORTED` for an opaque key. This is fragile and it doesn't stand a chance of being caught by static analysis. Please add a check directly in the main function in `psa_crypto.c` so that no matter what the driver wrapper does, the software implementation isn't invoked on an opaque key.

stevew817 commented on library/psa_crypto_driver_wrappers.c:
```diff
@@ -0,0 +1,378 @@
+                                           size_t *signature_length )
+{
+#if defined(PSA_CRYPTO_DRIVER_PRESENT)
+    /* Try dynamically-registered SE interface first */
```
I was thinking that having it first makes the function flow a little clearer, since catching these early allows for a cleaner switch case on the location value afterwards. Otherwise this block needs to move into the default case on the location switch. Is that what you're after?

gilles-peskine-arm replied:
Yes, I'm planning to put this block in the `default:` case in the generated code. It doesn't really matter now since this is only about microefficiency.

stevew817 replied:
Okay, I'll leave the change to you then.

stevew817 commented on library/psa_crypto_driver_wrappers.c:
```diff
@@ -0,0 +1,378 @@
+static psa_status_t get_expected_key_size( const psa_key_attributes_t *attributes,
+                                           size_t *expected_size )
+{
+    if( PSA_KEY_LIFETIME_GET_LOCATION( attributes->core.lifetime ) == PSA_KEY_LOCATION_LOCAL_STORAGE )
```
Looks like it. Thanks for pointing that out.

ronald-cron-arm commented on tests/suites/test_suite_psa_crypto_driver_wrappers.function:
```diff
@@ -0,0 +1,243 @@
+/* BEGIN_HEADER */
```
Thinking about it, it seems to me that we don't really need that. This is testing test code and the test code is PSA code or mbedtls code anyway (thus already tested) . I think it would be more useful to focus on amending/improving the PSA tests to check that when an operation (based on C pre-processor defines) is supposed to be done by a test driver we check that the test driver is called as expected (check of the call and check of the call parameters). If there is no equivalent test to the signature tests done here, it would be great to add them to the PSA tests.

gilles-peskine-arm replied:
I don't understand your comment: we don't really need what?

There are several uses for test drivers:

* Plug in transparent drivers that are expected to work normally and run the normal tests.
* Plug in opaque drivers that are expected to work normally and possibly have extra instrumentation (e.g. to verify that the driver was called when it should have), and run dedicated tests that support opaque drivers (which most don't).
* Plug in transparent drivers that are expected to work normally and have extra instrumentation and run dedicated tests that exercise the extra instrumentation.
* Plug in specially crafted drivers that can be instructed to misbehave in certain ways (e.g. returning an error), and check that the core handles the misbehavior correctly, from dedicated tests.

`test_suite_psa_crypto_driver_wrappers.function` gets started on the dedicated tests that will validate the wrapper code, validating properties such as making sure that the wrapper code calls the correct entry point.

ronald-cron-arm replied:
Sorry, I mixed up what was done in #3605 with what is done here. The tests below call the PSA API (not the test driver entry points as I was thinking when writing my comment above) and check the dispatch to the test driver which is indeed something we want to test. As I mentioned above this could be done by amending the PSA tests but this would probably make them too complex. To make it short, just forget about my comment above and sorry for the noise.

stevew817 replied:
Thanks. Can you close/resolve this then?

ronald-cron-arm replied:
Indeed.

stevew817 commented:
@gilles-peskine-arm and @ronald-cron-arm: pushed your requested changes
@bensze01 please re-approve.

stevew817 commented:
Rebased on top of development to get the fix for MBEDTLS_STATIC_TESTABLE in `check_names`

ronald-cron-arm commented on tests/suites/test_suite_psa_crypto_driver_wrappers.function:
```diff
@@ -0,0 +1,243 @@
+    psa_reset_key_attributes( &attributes );
+    psa_destroy_key( handle );
+    PSA_DONE( );
+    test_transparent_signature_sign_hash_status = PSA_ERROR_NOT_SUPPORTED;
```
With a structure grouping those variables, just reset the structure using the TEST_DRIVER_SIGNATURE_INIT macro probably.

stevew817 commented:
@gilles-peskine-arm Applied feedback about the internal function's visibility.

ronald-cron-arm commented on tests/src/drivers/keygen.c:
```diff
@@ -0,0 +1,130 @@
+#include <string.h>
+
+/* If non-null, on success, copy this to the output. */
+void *test_driver_keygen_forced_output = NULL;
```
With a structure containing those four variables please define a TEST_DRIVER_KEYGEN_INIT macro to initialize/reset the structure.

ronald-cron-arm commented on tests/src/drivers/signature.c:
```diff
@@ -0,0 +1,296 @@
+#include <string.h>
+
+/* If non-null, on success, copy this to the output. */
+void *test_driver_forced_output = NULL;
```
Ditto, use of an xyz_INIT macro.

gilles-peskine-arm reviewed (dismissed):
Thanks for the changes!

This is a partial approval: as a reviewer, I have not reviewed the unit tests (`tests/suites/test_suite_psa_crypto_driver_wrappers.*`). As a gatekeeper, I'll be willing to merge this pull request if at least one reviewer reviews the unit tests, in addition to a second approval for the rest of the PR.

ronald-cron-arm reviewed (changes requested):
I've reviewed the testing part. I have some proposals for improvement. Otherwise, concerning the reorganization of the psa_crypto.c code to minimize the amount of code copied from psa_crypto.c to the test drivers, I am eventually rather in favor of doing it in another PR as you proposed. There is already a lot of things in this PR (see comment thread in signature.c).

stevew817 commented on tests/suites/test_suite_psa_crypto_driver_wrappers.function:
```diff
@@ -0,0 +1,243 @@
+
+#include "drivers/test_driver.h"
+
+uint8_t test_secp256r1_key_data[32] = {
```
Sure.

stevew817 commented on library/psa_crypto.c:
```diff
@@ -3821,9 +3785,105 @@ psa_status_t psa_verify_hash( psa_key_handle_t handle,
+        goto exit;
+
+    /* If the operation was not supported by any accelerator, try fallback. */
+    status = psa_sign_hash_internal( slot,
```
Since we've reverted that for now, this comment is not relevant anymore?

ronald-cron-arm replied:
Not relevant for this PR just for future work in that direction (if any).

stevew817 commented on tests/include/drivers/signature.h:
```diff
@@ -0,0 +1,70 @@
+extern void *test_driver_forced_output;
+extern size_t test_driver_forced_output_length;
+
+extern psa_status_t test_transparent_signature_sign_hash_status;
```
Sure.

stevew817 commented:
@ronald-cron-arm I think I applied all of your feedback now. Please review.

gilles-peskine-arm commented on tests/include/drivers/keygen.h:
```diff
@@ -29,11 +29,25 @@
-
-extern psa_status_t test_transparent_keygen_status;
-extern unsigned long test_transparent_keygen_hit;
+struct test_driver_keygen_hooks_s {
```
Please define a type (`typedef … …_t`) and use it in preference to `struct xxx_s`. The fact that the type is a struct is only relevant when accessing its fields, not when referencing the type (e.g. when declaring a variable). Defining a struct name is optional: here I don't see the point.

(The reason `psa/crypto_struct.h` refers to `struct xxx_s` is that the type is defined in another file, which in turn is for a reason that doesn't apply here: the existence of the type is public and standard, but the struct layout is implementation-dependent.)

stevew817 replied:
That's how I set it up initially until I looked at `psa_structs.h`. Changed.

gilles-peskine-arm commented on tests/include/drivers/test_driver.h:
```diff
@@ -1,5 +1,5 @@
 /*
- * Umbrella include for all of the test driver functionality
+ * Umbrella include for all of the test driver functionalities
```
Minor: no, “functionality” is correct here. It's uncountable.

stevew817 replied:
Sure, changed back

gilles-peskine-arm commented on tests/suites/test_suite_psa_crypto_driver_wrappers.function:
```diff
@@ -0,0 +1,185 @@
+    psa_reset_key_attributes( &attributes );
+    psa_destroy_key( handle );
+    PSA_DONE( );
+    test_driver_signature_sign_hooks = test_driver_signature_hooks_init();
```
***Future*** note: we should define some function `mbedtls_test_driver_reset_all_instrumentation()` (name TBD) that we can use everywhere without having to care exactly which structures need resetting.

stevew817 replied:
Future notes might be better left in a ticket :)

gilles-peskine-arm reviewed (approved):
Thanks for the updates, again. As before, I haven't reviewed the unit tests in detail, only a cursory review, and I'm prepared to accept them with one code review. I've reviewed the rest of the pull request and am happy with it.

stevew817 commented on tests/include/drivers/signature.h:
```diff
@@ -0,0 +1,70 @@
+/*
```
The folder is already under `/tests/`. I think it's pretty clear these belong to whatever is inside that folder.
Simultaneously I think `tests/include/test/` is a bit redundant...

ronald-cron-arm commented on tests/include/drivers/keygen.h:
```diff
@@ -29,11 +29,25 @@
+    unsigned long hits;
+} test_driver_keygen_hooks_t;
+
+#define TEST_DRIVER_KEYGEN_INIT { NULL, 0, PSA_ERROR_NOT_SUPPORTED, 0 }
```
If you add a cast to the macro definition:
#define TEST_DRIVER_KEYGEN_INIT ( (test_driver_keygen_hooks_t){ NULL, 0, PSA_ERROR_NOT_SUPPORTED, 0 } ), you don't need the inline function. I think it would be better to do it that way as *_init() functions in the code base are used slightly differently.

gilles-peskine-arm replied:
This `init` function parallel to the `INIT` macro follows the PSA model. Going forward, it's likely that we'll move to this model in Mbed TLS. A function is better because it's typed. The only advantage of the macro is that you can use it as a static analyzer, whereas a function can't be used there.

ronald-cron-arm replied:
ok, fine by me then.

ronald-cron-arm reviewed (commented):
Thanks a lot for all the changes. Two comments left on my side.
First one here:
The idea of "include/test/drivers" and not "include/drivers" is to make things clearer when including header files. Currently we include them with "drivers/..." thus it is not clear that it is a test driver header from that perspective.

stevew817 commented:
@ronald-cron-arm changed directory structure as you requested.

ronald-cron-arm reviewed (approved):
Thanks again for all you work and rework on this PR. This looks good to me.
